staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_patch.go

```diff
@@ -0,0 +1,81 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package etcd3
+
+import (
+	"regexp"
+	"strings"
+
+	"github.com/kcp-dev/logicalcluster"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	genericapirequest "k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/klog/v2"
+)
+
+var (
+	// matches name/remainder, capturing name
+	wildcardClusterNameRegex = regexp.MustCompile(`^([^/]+)\/.+`)
+
+	// matches identity/name/remainder, capturing name
+	crdWildcardPartialMetadataClusterNameRegex = regexp.MustCompile(`^[^/]+\/([^/]+)\/.+`)
+)
+
+// adjustClusterNameIfWildcard determines the logical cluster name. If this is not a wildcard list/watch request,
+// the cluster name is returned unmodified. Otherwise, the cluster name is extracted from the key based on whether
+// it is a partial metadata request (prefix/identity/clusterName/remainder) or not (prefix/clusterName/remainder).
+func adjustClusterNameIfWildcard(cluster *genericapirequest.Cluster, crdRequest bool, keyPrefix, key string) logicalcluster.Name {
+	if cluster.Name != logicalcluster.Wildcard {
+		return cluster.Name
+	}
+
+	sub := strings.TrimPrefix(key, keyPrefix)
+
+	var regex *regexp.Regexp
+
+	if cluster.PartialMetadataRequest && crdRequest {
+		regex = crdWildcardPartialMetadataClusterNameRegex
+	} else {
+		regex = wildcardClusterNameRegex
+	}
+
+	matches := regex.FindStringSubmatch(sub)
+	if len(matches) >= 2 {
+		return logicalcluster.New(matches[1])
+	}
+
+	return logicalcluster.Name{}
+}
+
+func setClusterNameOnDecodedObject(obj interface{}, clusterName logicalcluster.Name) {
+	if clusterName.Empty() {
+		klog.Errorf("Got empty clusterName!")
```

Comment:
do we care? Why not blindly write it?

Reply:
This came along when I was shuffling code around. I think before my PR, it was something to look out for, in case we unexpectedly didn't have a cluster name. Happy to remove this